### ISSUE TITLE
Suggestion: readouts per second counter

### ISSUE THREAD

**Issue by nightsfan** [sanderland/katrain#607]
# I have a feature request

I think it would be helpful for comparing the speed of different networks if there were a (small) readouts per second counter somewhere on the interface (like there was in Lizzie).

Currently, I can see the numbers going up on the individual coordinates, but it's hard to get an overall picture of relative speed (e.g. comparing the new 18-block network with the 40-block one).

**Comment by sanderland** [comment]
The ability to fire off and deal with many analysis requests in parallel makes this not so easy to define, and I'm not sure I'd want to clutter the UI with it. If someone contributes this, I'd happily accept as console/debug output of the engine to help those comparing speeds though.

**Comment by kylebakerio** [comment]
Just want to say, this would be helpful when attempting to tune katago locally. Specifically, I've been trying to, and while the official advice has been "just go with what benchmark/genconfig suggests", I find (for example) that loading in an SGF, I see a far slower time to complete board analysis than if I massively up the search threads, analysis threads, and maxbatchsize. (I have 16gb of vram, so I was trying extremely high numbers, like 6144 batchsize, and hundreds of search/analysis threads.)

On the discord, though, they discourage this, saying you get a small ELO penalty for every thread (and suggested that maybe it would break the MCST altogether, though this wasn't explained and seemed to be speculative; felt like no one there had tried using it this way honestly)--but katago has plenty of ELO to spare, and I'm a lowly SDK, so the tradeoff seems worth it to me, especially since that ELO drop is "given the same time constraint".

I'd have to actually run some tests to verify that it doesn't totally "break" katago to be used that way, but it would be nice to actually see playouts per second. It seems those numbers could be inferred from currently visible data--aren't we saying playouts for each move on a given position listed already? Can't we just sum those somewhere? And then show them next to the engine status dot?

related, but maybe separate issue, let me know if I should discuss elsewhere instead:  it looks to me like katrain includes a config setting for 12 analysis threads out of the box. I'm kind of curious about that setting. After my in depth discussions on katago discord, my machine seems to like 16-24 threads (but higher aren't even tested, because benchmark is biased towards highest ELO....), and reading the example analysis_example.cfg file seems to suggest that for the kind of game-analysis mode I'm most interested in, mroe analysis threads and fewer search threads is the right approach. So, going with 24 threads as per recommendation in discord, and following guidelines in that example, I am doing 6 analysis threads and 4 search threads (6*4 = 24), with batchsize shrunk down to match that number.

Is 12 hardcoded, or set during initial load  (when I presume there's a genconfig step?) of a model?

**Comment by sanderland** [comment]
numSearchThreads = 8 is in analysis_config.cfg. This is a rather technical detail of MCTS search, in that you don't know the next node to evaluate while some other threads are running, which leads to the strength hit.
analysisthreads = 12 is in the config  (= the number of simultaneous queries to evaluate, which mainly hits when analyzing entire games/all moves)
there are no GUI option to change them, as they're very technical, and I personally don't change them.

**Comment by kylebakerio** [comment]
That 12*8 _seems_ pretty high to me, according to what I've gleaned from diving into this and asking probably too many questions on discord and reading the docs. What's the default maxBatchSize? Should be at least 12*8=96, right? Actually, now that I type that, I recall that's what it was.

When benchmarking, I see (for my gpu, a laptop with an 80w 3080) optimal batch sizes of 16-24.

To be fair, I saw much faster analysis of the game score/win% graph when I had huge numbers of threads, but the suggestion was unanimously "use the benchmark settings" and when I pushed for details on analysis threads, "still use the benchmark saturation batch size limit to decide, and set it so that analysis * search thread = benchmark search thread number".

I've just sent a message there asking how I can run elo tests locally on this kind of experiment. If it _does_ produce reasonable results, maybe a setting to run for analysis vs. strength would be ideal, and maybe a local benchmarking step to see how many analysis threads can be run effectively (could perhaps set a required minimum elo, other limiting factor is vram)? If nothing else, it might make sense to add a gui option for them.

On the other hand, if the consensus there is correct, maybe the current setting is really not ideal / already too high? Supposedly the setting out of the box in katrain should actually hurt katago's elo rating on most consumer hardware if I understand correctly, if people are focused on playing the strongest possible katago their machine can support.

**Comment by kylebakerio** [comment]
Of note, someone else recently pointed out that changing the "time" parameter will change the tune significantly, and can allow more readouts; and considering I'm usually allowing thousands (to tens/hundreds of thousands, rarely even a million or 10million if I step away while it's checking), there's room for a lot more threads.

However, more playouts become "necessary" for confidence, I guess; so, the 'full board deeper analysis' kind of becomes essential. In theory, I suppose one could compensate for that by also selecting the number of 'fast readouts' (iirc) to match whatever this safe number would be so that initial games get fully read deeply enough by default.
